manual/manual/cmds/instrumented-runtime.etex

```diff
@@ -0,0 +1,312 @@
+current directory.
+
+\subsubsection*{sss:instr-runtime-build-more}{More build examples}
+
+When using the {\em dune} build system, this compiler invocation can be
+replicated using the {\tt flags} {\tt stanza} when building an executable.
+
+\begin{verbatim}
+       (executable
+         (name program)
+         (flags "-runtime-variant=i"))
+\end{verbatim}
+
+The instrumented runtime can as well be used with OCaml's bytecode interpreter.
+This can be done by either compiling the program using the
+{\tt -runtime-variant=i} flag with {\tt ocamlc}, or by running the generated
+bytecode through {\tt ocamlruni}:
+
+\begin{verbatim}
+       ocamlc program.ml -o program.byte
+       OCAML_EVENTLOG_ENABLED=1 ocamlruni program.byte
+\end{verbatim}
+
+See chapter~\ref{c:camlc} and chapter~\ref{c:runtime} for more information about
+{\tt ocamlc} and {\tt ocamlrun}.
+
+\subsection{ss:instr-runtime-prefix}{Trace filename}
+
+The default trace filename is {\tt caml-\{pid\}.eventlog}, where {\tt \{pid\}}
+is the process identifier of the traced program.
+
+This filename can also be specified using the
+{\tt OCAML\_EVENTLOG\_PREFIX} environment variable.
+The given path will be suffixed with {\tt \{.pid\}.eventlog}.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_PREFIX=/tmp/a_prefix OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+Note that this will only affect the prefix of the trace file, there is no
+option to specify the full effective file name.
+This restriction is in place to make room for future improvements to the
+instrumented runtime, where the single trace file per session design
+may replaced.
+
+In this example, the trace will be available at path
+{\tt /tmp/a_prefix.\{pid\}.eventlog}.
+
+For scripting purpose, matching against `\{pid\}` as well as the
+{\tt .eventlog} file extension should provide enough control over
+the generated files.
+
+\subsection{ss:instr-runtime-pause}{Pausing and resuming tracing}
+Primitives are available to control events collection at runtime.
+
+{\tt OCAML\_EVENTLOG\_ENABLED} can be set to the {\tt p} flag in order
+to {\em pause} events collection after the program has started.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=p ./program
+\end{verbatim}
+
+The program will have to resume events collection explicitly.
+This can be done via {\tt Gc.eventlog_resume} and
+{\tt Gc.eventlog_pause}, available in the {\stdmoduleref{Gc}} module.
+
+Running the program provided earlier with {\tt OCAML\_EVENTLOG\_ENABLED=p}
+will for example yield the following result.
+
+\begin{verbatim}
+        $ OCAML_EVENTLOG_ENABLED=p ./a.out
+        $ ocaml-eventlog-report caml-{pid}.eventlog
+        ==== eventlog/flush
+        median flush time: 58ns
+        total flush time: 58ns
+        flush count: 1
+\end{verbatim}
+
+The resulting trace contains only one event payload, namely a {\em flush} event,
+indicating how much time was taken to flush the trace file to disk.
```

Comment:
This is not a documentation issue but I find it strange that a paused event log that remains so does anything to the program. I would expect the file to be created but for it to be empty. (I can see however that you'd have to detect "was never unpaused in the program" condition that it may not be suitable to do so for perf reasons)

Reply:
I think it is sane to have at least the CTF header being written, as a part of the file opening process.
The flush event however could be avoided, without trickery, its intended purpose is to time the execution of the eventlog flush function, however at least one eventlog flush happens in any instrumented program: atexit, which forces a buffer flush.
This buffer could very well be empty however (as is the case for a `paused` program), so this extra flushing information could be skipped, since the flush function will exactly flush nothing. (and not introduce any performance penalty whatsoever)